### PR TITLE
Mpoly interface

### PR DESCRIPTION
**Pull request by wbhart**
This PR extends the fmpq_mpoly interface and renames a number of functions to be in line with those for generics in AbstractAlgebra. It also adds the evaluate function, which now exists in Flint.

However, I am not happy with this in a number of respects and I think we should discuss some issues before rolling this out for fmpz_mpoly and nmod_mpoly.

* Currently if you want to write an exponent or coefficient beyond the current length of a poly, you have to call fit! just like you do for the generic code. But none of the other Flint polynomial modules (fmpz_poly, fmpq_poly, etc.) implement fit!, as it isn't generally needed.

* This interface that allows changing monomials and coefficients is low level, but allows the user to create polynomials that don't make any sense. I'm beginning to think this is not something we should allow for the Flint interface. They only reason we allow it for generics is because we need to use it internally. On the other hand, if we want any of the generic code to also work for Flint polynomials, this internal interface will be needed I guess.

* However, the question is whether we should only support adding terms immediately at the end of the polynomial (push_term!) and only if the new term is in the correct order. Perhaps all the generic functions we want to implement efficiently (such as derivative, integral, change_ring and the like) can be implemented efficiently in terms of this one safe function. Or perhaps we just require the implementer of those functions to create an array of coefficients and an array of exponent vectors (which can contain duplicates and out-of-order terms) and do a single safe construction of the polynomial from those.

In other words, do we really need a function to allow modification of an exponent vector, or to allow setting a coefficient to zero or to set coefficients and exponents of terms well beyond the end of the polynomial? Quite possibly these are not required for efficient implementation of generic functions that are also designed to work with Flint.

### CODE REVIEW

**Comment by wbhart** [comment]
Another thing to consider is that it might be hard to get the interface for constructing a polynomial from an array of coefficients and an array of exponent vectors the same between the generic code and the Flint code, since Flint can't really do anything with a two dimensional array of exponents, whereas this is what is required for the generic polynomials in AbstractAlgebra. So writing generic code in terms of this primitive might not be very efficient. It's worse if we consider that the array of coefficients might be supplied by the user as an array of Ints instead of an array of whatever type the coefficients need to be. So copying of arrays is going to happen all over the place.

The push_term! interface would discourage creating an array of Int's for the coefficients or a double array of UInts for the exponents that just needs to be copied into the actual polynomial structure on creation.

I guess the question that will decide it is if there is a generic function that we want to implement for generic polynomials and Flint polynomials for which the terms of the result cannot be written in order. It would also have to be time critical, which probably means it would have to have linear time in the number of terms. But all of these seem to be writing the terms of their result in order, which means that the push_term! interface would be sufficient.

**Comment by thofma** [comment]
I think we should not expose all the low level exponent/term modifying functions as part of the interface.

Requiring the efficient construction to use arrays of coefficients and exponents (not necessarily "cleaned up") sounds like a good idea to me. But keep in mind that I am not an export for multivariate polynomials and I don't intend to become one.

**Comment by thofma** [comment]
By the way, do you intend to bump the flint version here? Then I need to build new binaries before this PR here passes on Windows.

**Comment by wbhart** [comment]
I do intend to update the Flint version, as we need Daniel's latest code to fix some bugs, and for evaluation and for the latest interface. He's probably going to rename some functions too, so it'll need another bump later too.

**Comment by thofma** [comment]
I have rebuild the binaries.

**Comment by wbhart** [comment]
Thanks. I will see if I can finalise the interface over the next day or so.

**Comment by wbhart** [comment]
I think I have decided what to do. Basically there were three options:

* Introduce a push_term! function which allows appending a term so long as it is in the correct order with respect to the existing terms. Allow existing coefficients to be written so long as they are not set to zero, using setcoeff!

* Allow writing coefficients and exponents as required, as things are now, risking inconsistent polynomials

* Only support creating polynomials from an array of coefficients and an array of exponent vectors, always sorting, combining terms with the same exponent vector and removing zero terms.

The first option has the disadvantage of needing to check whether the term can be validly inserted at the end of the polynomial, i.e. that the ordering is respected. This would mean writing functions to check the ordering for all Flint multivariate polynomial types. It also has the disadvantage of not being very flexible.

The second option risks inconsistent polynomials being created, but is the most efficient and flexible option.

The third option is difficult to do in an efficient way. Currently we can create generic polys with an array of coefficients and a 2d array of exponents. Flint polynomials on the other hand are created from an array of coefficients and an array of coefficient vectors. The latter is the easiest for the user to use, but the former is the most efficient for the generic implementation; it is used extensively internally in the implementation of the functionality in MPoly.jl.

I've decided that the push_term! option is the least attractive, and so I won't implement that.

The second option is efficient and flexible, but it should only be used internally, not by the user. It's already implemented, so nothing to change there. We just won't put it in the user docs. But functions like derivative and change_ring can use this. However, I think I will get rid of the fit! function for Flint polynomials and put the fit! inside the set_exponent_vector! and setcoeff! functions, so that it's most similar to the existing Flint polynomial rings.

However, the third option is an attractive option for the user, if they want to efficiently create polynomials. The only implementation missing is one that creates generic polynomials from a list of coefficients and an array of exponent vectors. I will implement this and we'll document that as the recommended way for users to create polynomials. There's a small overhead in copying the array, and there's considerable overhead for large polynomials due to allocating a lot of small arrays. But I think it is the most sensible option for the user. If this becomes a big bottleneck for them, they can poke around in the library code and find the setcoeff! and set_exponent_vector! functions.

I think this strategy will give the greatest performance/safety ratio. Sparse multivariates are somehow a worst case for CAS's in that they can take a lot of time to create, compared to the operations you do on them.

**Comment by thofma** [comment]
This sounds like a reasonable approach. I agree with your conclusion.

**Comment by wbhart** [comment]
This should be right to go now @thofma

**Comment by thofma** [comment]
Thanks.
